File: 24b1291_week4 Atharva/Find the smallest Divisor.cpp
class Solution {
public:
    int smallestDivisor(vector<int>& nums, int t) {
        int left = 1, right = *max_element(nums.begin(), nums.end());

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int sum = 0;

            for (int num : nums) {
                sum += (num + mid - 1) / mid; 
            }

            if (sum <= t) {
                right = mid - 1; 
            } else {
                left = mid + 1; 
            }
        }

        return left;
    }
};

File: Chavda_23B0368/Week_4/Find_the_smallest_divisor.java
package Week_4;

class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int n = nums.length;
        if(n > threshold)
            return -1;

        int start = 1, end = 1;
        for(int num : nums) {
            end = Math.max(end, num);
        }

        while(start <= end) {
            int mid = end - (end-start)/2;
            if(sum(nums, mid) <= threshold)
                end = mid - 1;
            else 
                start = mid + 1;
        }

        return start;
    }

    private int sum(int[] nums, int div) {
        int sum = 0;
        for(int num : nums) {
            if(num%div == 0)
                sum += (num/div);
            else 
                sum += (num/div) + 1;
        }
        return sum;
    }
}